crates/driver/src/domain/competition/mod.rs

```diff
@@ -328,6 +302,127 @@ impl Competition {
     }
 }
 
+/// Encode the solutions into settlements.
+async fn encode_solutions(
+    auction: Auction,
+    eth: Ethereum,
+    simulator: Simulator,
+    solutions: Vec<Solution>,
+) -> Vec<(solution::Id, Result<Settlement, solution::Error>)> {
+    let (sender, mut receiver) = tokio::sync::mpsc::unbounded_channel();
+    let auction = Arc::new(auction);
+    let eth = Arc::new(eth);
+    let simulator = Arc::new(simulator);
+
+    let futures = solutions
+        .into_iter()
+        .map(|solution| {
+            let sender = sender.clone();
+            let auction = auction.clone();
+            let eth = eth.clone();
+            let simulator = simulator.clone();
+            async move {
+                let id = solution.id();
+                observe::encoding(id);
+                let settlement = solution.encode(&auction, &eth, &simulator).await;
+                let _ = sender.send((id, settlement));
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let deadline = auction.deadline().remaining().unwrap_or_default();
+    if tokio::time::timeout(deadline, tokio::spawn(join_all(futures)))
+        .await
+        .is_err()
+    {
+        tracing::warn!("reached timeout while encoding");
```

Comment:
Aren't these events that the observe crate should handle?

Reply:
I'm definitely in favor of phasing out the `observe` crate as it has several downsides compared to simply logging wherever the event happened.

Reply:
It wraps the logging and metrics which can be quite cumbersome into one-liner, which I like. What would be the downsides?

Reply:
I wrote down a little rant a while back:
1. AFAICS the only thing that is considered an improvement is that all the logs are in one or very few places. This supposedly helps with identifying which logs are really necessary and which ones aren’t and should help us with being critical about what’s being logged. I would say this is nothing a simple search for “tracing” can’t also achieve. You might say with just grepping you lose the connecting of which log is related or close to another log. However, the same problem applies with the hook based approach because with enough hooks you will also not know how hooks relate to each other.
2. tracing spans vs. passing around values: The reductionistic argument of “global state vs. local state” does not make sense IMO. First of all logging in another module means you will either leak types that otherwise would not have been leaked or not log something. And secondly putting state in tracing spans is IMO “less”. global (and way more convenient) than passing the values around manually. Values that you pass around manually pollute the namespace and structs they get passed into whereas state within tracing spans is only accessible by their child tracing spans effectively meaning that from our perspective no state is global. You could make the argument that it’s “hidden” state but IMO it doesn’t hurt because it does not increase any complexity.
3. logging code locations: Logs using tracing will print in which module the log originated from. This is great when you want all logs of a certain code section (e.g. price competition). With the new approach you will have to know which exact log messages happen in the code section you are interested in and have to build kibana queries like log: "first log" or log: "second log"…. This is inconvenient compared to log: "general module". What I find especially frustrating about this short coming is that the new driver is architected very hierarchical which is even more compatible with this idea than the remaining code base which is partly scattered around.
4. It’s a lot more overhead. With the new approach you have to first create a hook, possibly export some type and pass around more state just to get a log message. This could lead to us writing fewer logs which is unfortunate because the amount of logging we have atm can already be somewhat lacking. Also without tracing spans and module being printed automatically we have to think more ahead (and remember to add the boiler plate) to arrive at logs that can be efficiently analysed.
5. Temporarily running the driver with different logs levels for different modules is also more difficult when there is only 1 module. To work around that you would have to set up sub modules. All of this boiler plate is only needed to get a IMO very questionable advantage.

Reply:
I think 3 and 5 are strong practical reasons to phase out this approach (however not in this PR and I prefer having consistency over slowly starting to go down a route we prefer).

I did like that the domain code doesn't have to worry about what "observe" means (ie. do we just want to log or record a metric or do other observation things) and it reads quite nicely. 
That being said, this could probably also be achieved by making the observe module local to each domain module. Then there is still the boiler-plate arguments, which in my mind is a trade-off we were willing to accept for DDD (we have a lot of boilerplate model conversion code).